tsdb/db.go

```diff
@@ -1522,34 +1522,40 @@ func (db *DB) CleanTombstones() (err error) {
 	start := time.Now()
 	defer db.metrics.tombCleanTimer.Observe(time.Since(start).Seconds())
 
-	newUIDs := []ulid.ULID{}
-	defer func() {
-		// If any error is caused, we need to delete all the new directory created.
-		if err != nil {
-			for _, uid := range newUIDs {
+	cleanUpCompleted := false
+	// Repeat cleanup until there is no tombstones left.
+	for !cleanUpCompleted {
+		cleanUpCompleted = true
+
+		for _, pb := range db.Blocks() {
+			uid, safeToDelete, cleanErr := pb.CleanTombstones(db.Dir(), db.compactor)
+			if cleanErr != nil {
+				return errors.Wrapf(cleanErr, "clean tombstones: %s", pb.Dir())
+			}
+			if !safeToDelete {
+				// There was nothing to clean.
+				continue
+			}
+
+			// In case tombstones of the old block covers the whole block,
+			// then there would be no resultant block to tell the parent.
+			pb.meta.Compaction.Deletable = safeToDelete
+			cleanUpCompleted = false
+			if err = db.reload(); err == nil { // Will try to delete old block.
+				// Successful reload will change the existing blocks.
+				// We need to loop over the new set of blocks.
+				break
+			}
+
+			// Delete new block if it was created.
+			if uid != nil {
```

Comment:
Feels like there could be a race here since we only know the reload failed. It is also possible the time/size retention kicked in and deletion of those block failed and the one we wanted to delete was deleted. So we should reliably verify that the parent still exists else we lose data. Maybe call `db.Blocks()` here and see if the parent still exists and then delete.

Can you also add a test case for this?

Reply:
If time/size retention kicks in and deletes the parent in the middle of cleanTombstones(), I don't see a problem in cleaning the new block as well since they also should be cleaned the retention limits, don't they?

Reply:
Shouldn't we already be holding a lock so only one compaction-y event can be ongoing at once?

Reply:
There is a lock over here already:
https://github.com/prometheus/prometheus/blob/e483cff61ff1b68e8925db1059393c4d36af9df5/tsdb/db.go#L1034-L1037

But yeah, I think it is possible to have a race here where cleanTombstones and a deletion cause by retention limits can happen at the same time:
https://github.com/prometheus/prometheus/blob/e483cff61ff1b68e8925db1059393c4d36af9df5/tsdb/db.go#L1053-L1055
There is no problem on Linux with this, but I think that windows would complain if we try to delete a file that does not exist anymore.

Would it be safe to move the unlock from line 1037 to line 1056? We would lock this whole deletion process.

Reply:
That's a bug then, we shouldn't be doing operations like that in parallel.

Reply:
Does the last commit solve the race?

Reply:
`uid != nil` does not signify a non-empty block, it should be `uid != nil && *uid != ulid.ULID{}` since thats what compactor returns.